viewer/src/Main.rs

```diff
@@ -0,0 +1,324 @@
+                            } //minus
+                            Keycode::E => {
+                                self.scale =
+                                    (self.max_stop - self.min_start) / (self.window_width as u64)
+                            } //reset
+                            Keycode::A => self.scroll = self.scroll.saturating_add(keycount),
+                            Keycode::S => self.scroll = self.scroll.saturating_sub(keycount),
+                            Keycode::D => self.scroll = 0,
+                            _ => {}
+                        }
+                    }
+                    Event::MouseButtonDown { x, y, .. } => {
+                        for zone in &self.draw_zones {
+                            if zone.x_start < x.try_into().unwrap()
+                                && zone.x_stop > x.try_into().unwrap()
+                                && zone.y_start < y.try_into().unwrap()
+                                && zone.y_stop > y.try_into().unwrap()
+                            {
+                                draw_x = x;
+                                draw_y = y;
+                                draw_tag = zone.tag_data.0;
+                                draw_len = zone.tag_data.2 - zone.tag_data.1;
+                            }
+                        }
+                    }
+                    Event::MouseButtonUp { .. } => {
+                        draw_x = 0;
+                        draw_y = 0;
+                    }
+                    _ => {}
+                }
+            }
+            if prev_keycount == keycount {
+                keycount = 0;
+            }
+            prev_keycount = keycount;
+
+            self.canvas.set_draw_color(self.background_color);
+            self.canvas.clear();
+            self.draw_zones.clear();
+            for span in &spans {
+                self.draw_span(span);
+            }
+            if (draw_x > 0) && (draw_y > 0) {
+                Self::draw_text(
+                    &mut self.canvas,
+                    &self.texture_creator,
+                    &font,
+                    draw_x,
+                    draw_y,
+                    draw_tag.try_into().unwrap(),
+                    draw_len.try_into().unwrap(),
+                )?;
+            }
+            self.canvas.present();
+            thread::sleep(Duration::new(0, 1_000_000_000u32 / 30));
+        }
+
+        Ok(())
+    }
+}
+
+pub fn load_args() -> Vec<Span> {
+    let mut spans = vec![];
+    let args: Vec<_> = env::args().collect();
+    match args.len() {
+        1 => {panic!("Command line arguments were not provided. Format: (file path) (span range start) (span range stop) (tag range start) (tag range stop).")},
+        6 =>{
+            let mut file = File::open(&args[1]).expect("failed to open file");
+            let mut buffer = [0; 24];
+
+            for i in 0..args[3].parse::<u64>().unwrap(){
```

Comment:
I try to avoid bare unwrap() in favor of expect where possible.  These ones seem especially likely to trigger if someone misunderstands the format of the command line arguments, they'd probably appreciate a more useful panic message